BlueprintUI/Sources/Environment/Environment.swift

```diff
@@ -34,33 +34,111 @@ import UIKit
 ///             }
 ///         }
 ///     }
-public struct Environment {
+public struct Environment: Equatable {
+
+    typealias OnDidRead = (StorageKey) -> Void
+
+    enum LayoutPass: Equatable {
+        case measurement
+        case layout
+    }
 
     /// A default "empty" environment, with no values overridden.
     /// Each key will return its default value.
-    public static let empty = Environment()
+    public static var empty: Environment {
+        .init()
+    }
+
+    var readNotificationsEnabled: Bool = true
+
+    /// If the `Environment` contains any values.
+    var isEmpty: Bool {
+        values.isEmpty
+    }
+
+    private var values: [StorageKey: ValueBox] = [:]
+
+    private var measurementDidRead: OnDidRead?
+    private var layoutDidRead: [OnDidRead] = []
 
-    private var values: [ObjectIdentifier: Any] = [:]
+    private var onDidReadSubscriptions: [OnDidRead] {
```

Comment:
Nit: Rather than building an array, I'd just call the two sequentially inline in the code; since there's a small but non-zero cost to building this array.